src/Context/Traits/UserAwareContextTrait.php

```diff
@@ -84,31 +85,134 @@ public function loggedIn()
+     *             @type int $id User ID.
+     *             @type string $slug User slug (nicename).
      *         }
      */
     public function createUser($userLogin, $userEmail, $args = [])
     {
         $args['user_email'] = $userEmail;
         $args['user_login'] = $userLogin;
 
-        $user = $this->getDriver()->user->create($args);
+        try {
+            $user = $this->getDriver()->user->create($args);
+        } catch (UnexpectedValueException $exception) {
+            $user = $this->getExistingMatchingUser($args);
+        }
 
-        return array(
+        $return_array = array(
             'id'   => $user->ID,
             'slug' => $user->user_nicename
         );
+
+        return $return_array;
+    }
+
+    /**
+     * Returns a user if all the passed parameters match
+     * throws an UnexpectedValueException if not
+     *
+     * if there is return a user ID
+     *
+     * @throws \UnexpectedValueException
+     * @return \WP_User $user
+     */
+    private function getExistingMatchingUser($args)
+    {
+        $user_id = $this->getUserIdFromLogin($args['user_login']);
+        $user = $this->getDriver()->user->get($user_id);
+
+        /* users can have more than one role so needs to be a special case */
+        if (array_key_exists('role', $args)) {
+            $this->checkUserHasRole($user, $args['role']);
+        }
+
+        /*
+         * Loop through each of the passed arguements.
+         * if they are arguments which apply to users
+         * then check that that which exist matches that which was specified.
+         */
+        foreach ($args as $parameter => $value) {
+            if ($parameter == 'password') {
+                try {
+                    if (! $this->getDriver()->user->validateCredentials($args['user_login'], $value)) {
+                        throw new \UnexpectedValueException('User with login : ' . $user->user_login . ' exists but password is incorrect');
+                    }
+                } catch (UnsupportedDriverActionException $exception) {
+                    // WPCLI can't do this yet.
+                }
+            }
+            if ($this->isValidUserParameter($parameter) && $user->$parameter != $args[$parameter]) {
+                throw new \UnexpectedValueException('User with login : ' . $user->user_login . 'exists, but ' . $parameter . ' is ' . $user->$parameter . ' not ' . $args[$parameter] . 'which was specified');
+            }
+        }
+
+        return $user;
+    }
+
+    /**
+     * Checks to see if the user has an assigned role or not.
+     *
+     * @param \WP_User $user
+     * @param string $role
+     *
+     * @throws \UnexpectedValueException
+     *
+     * @return boolean $retval True if the role does apply to the user.
+     */
+    private function checkUserHasRole($user, $role)
+    {
+        // $user->role can either be a string with 1 role in it or an array of roles.
+        if (is_string($user->roles)) {
```

Comment:
I think if we cast roles to an array, then we can remove this first clause and just use inarray.

Reply:
oh that's a much better idea 👍